dart/dynamics/InverseKinematics.hpp

```diff
@@ -130,11 +130,93 @@ class InverseKinematics : public common::Subject
+  /// findSolution() instead.
+  DART_DEPRECATED(6.8)
+  bool solve(Eigen::VectorXd& positions, bool applySolution = true);
+
+  /// Finds a solution of the IK problem without applying the solution.
+  ///
+  /// The initial guess for the IK optimization problem is the current joint
+  /// positions of the target system. If the iterative solver fails to find a
+  /// successive solution, it attempts more to solve the problem with other seed
+  /// configurations or random configurations if enough seed is not provided.
+  ///
+  /// Here is the pseudocode as described above:
+  ///
+  /// \code
+  /// attempts <- 0
+  /// initial_guess <- current_joint_positions
+  /// while attempts <= max_attempts:
+  ///   result <- solve(initial_guess)
+  ///   if result = success:
+  ///     return
+  ///   else:
+  ///     attempts <- attempts + 1
+  ///     if attempts <= num_seed:
+  ///       initial_guess <- seed[attempts - 1]
+  ///     else:
+  ///       initial_guess <- random_configuration  // within the bounds
+  /// \endcode
+  ///
+  /// By default, the max_attempts is 1, but this can be changed by calling
+  /// InverseKinematics::getSolver() and casting the SolverPtr to an
+  /// optimizer::GradientDescentSolver (unless you have changed the Solver type)
+  /// and then calling GradientDescentSolver::setMaxAttempts(std::size_t).
+  ///
+  /// By default, the list of seeds is empty, but they can be added by calling
+  /// InverseKinematics::getProblem() and then using
+  /// Problem::addSeed(Eigen::VectorXd).
+  ///
+  /// Calling this function will automatically call Position::setLowerBounds(~)
+  /// and Position::setUpperBounds(~) with the lower/upper position bounds of
+  /// the corresponding Degrees Of Freedom. Problem::setDimension(~) will be
+  /// taken care of automatically, and Problem::setInitialGuess(~) will be
+  /// called with the current positions of the Degrees Of Freedom.
+  ///
+  /// \param[out] position The solution of the IK problem. If the solver failed
+  /// to find a solution then it will still set the position with the best
+  /// guess. For example, iterative solvers will fill \c positions with the last
+  /// result of the iterations.
+  /// \return True if a solution is successfully found.
+  /// \sa solveAndApply()
+  bool findSolution(Eigen::VectorXd& positions);
+
+  /// Identical to findSolution(), but this function applies the solution when
+  /// the solver successfully found a solution or \c allowIncompleteResult is
+  /// set to true.
+  ///
+  /// \param[in] allowIncompleteResult Allow to apply the solution even when
+  /// the solver failed to find solution. This option would be useful when an
+  /// iterative solver is used because they will often do a decent job of
+  /// getting a result close to a solution even if it failed to find the
+  /// solution.
+  /// \return True if a solution is successfully found
+  bool solveAndApply(bool allowIncompleteResult = false);
```

Comment:
I have mixed feelings on whether this should default to true or false. I understand why someone might exercise caution and avoid applying "unsolved" results, but for an iterative solver it's still better to hang onto the last unsolved result because if you try to run the solver again you'll be in a better position to find the solution than if you restart from the beginning. This is often done in order to solve an IK over the course of several GUI refreshes instead of needing to solve from scratch every rendering cycle.

Also, if the usual behavior for this function is to not apply the partial solution, then the user won't know where the configuration ended up if it failed to solve, and the user is left only knowing that a solution simply wasn't found.

Since we now have `findSolution(~)` which already defaults to not applying the result, I would recommend that `solveAndApply(~)` should default to applying the solution.

But I suppose this is a difference in preferences that mostly boils down to a difference in behavior between iterative solvers and closed-form solver (and we're trying to opaquely support both here), so I'm fine with keeping this as it is if there's a strong preference to have it this way.

Reply:
> Since we now have `findSolution(~)` which already defaults to not applying the result, I would recommend that `solveAndApply(~)` should default to applying the solution.

I'm also in two minds, but I think this statement made me to incline to defaulting the option to be true.

Maybe we could introduce `solveAndApplyOnSuccess()` to be more clear, but it feels like too much. But if you think it sounds good then I'm okay to add it.

Reply:
I think `solveAndApplyOnSuccess()` would be overcompensating on explicitness. I think a well-named function argument is preferable to a more explicit function name, because arguments make it easier to programmatically determine behavior, for example

```
ik->solveAndApply(allowIncomplete);
```

vs

```
if(allowIncomplete) { 
  ik->solveAndApplyOnSuccess();
} else { 
  ik->solveAndAlwaysApply();
}
```

Reply:
Agreed. Let me change the default value to true.